azpubsub/contextvalidator/src/main/scala/kafka/security/auth/AzPubSubAclAuthorizer.scala

```diff
@@ -0,0 +1,459 @@
+
+      acl.permissionType == permissionType &&
+        (acl.principal == principal
+          || (principal.getPrincipalType == KafkaPrincipal.USER_TYPE && acl.principal == KafkaPrincipal.WildCardUserTypePrincipal)
+          || (principal.getPrincipalType == KafkaPrincipal.ROLE_TYPE && acl.principal == KafkaPrincipal.WildCardRoleTypePrincipal ) ) &&
+        (operations == acl.operation || acl.operation == All) &&
+        (acl.host == host || acl.host == Acl.WildCardHost)
+    }.exists {
+
+      acl =>
+        LOGGER.debug(s"operation = $operations on resource = $resource from host = $host is $permissionType based on acl = $acl")
+
+        true
+    }
+  }
+
+  object AclChangedNotificationHandler extends NotificationHandler {
+    override def processNotification(notificationMessage: Array[Byte]) {
+      val resource: Resource = Resource.fromString(new String(notificationMessage, StandardCharsets.UTF_8))
+      inWriteLock(lock) {
+        val versionedAcls = getAclsFromZk(resource)
+        updateCache(resource, versionedAcls)
+      }
+    }
+  }
+
+  override def addAcls(acls: Set[Acl], resource: Resource) {
+    if (acls != null && acls.nonEmpty) {
+      inWriteLock(lock) {
+        updateResourceAcls(resource) { currentAcls =>
+          currentAcls ++ acls
+        }
+      }
+    }
+  }
+
+  override def removeAcls(aclsTobeRemoved: Set[Acl], resource: Resource): Boolean = {
+    inWriteLock(lock) {
+      updateResourceAcls(resource) { currentAcls =>
+        currentAcls -- aclsTobeRemoved
+      }
+    }
+  }
+
+  override def removeAcls(resource: Resource): Boolean = {
+    inWriteLock(lock) {
+      val result = zkClient.deleteResource(resource)
+      updateCache(resource, VersionedAcls(Set(), 0))
+      updateAclChangedFlag(resource)
+      result
+    }
+  }
+
+  override def getAcls(resource: Resource): Set[Acl] = {
+    inReadLock(lock) {
+      aclCache.get(resource).map(_.acls).getOrElse(Set.empty[Acl])
+    }
+  }
+
+  override def getAcls(principal: KafkaPrincipal): Map[Resource, Set[Acl]] = {
+    inReadLock(lock) {
+      aclCache.mapValues { versionedAcls =>
+        versionedAcls.acls.filter(_.principal == principal)
+      }.filter { case (_, acls) =>
+        acls.nonEmpty
+      }.toMap
+    }
+  }
+
+  override def getAcls(): Map[Resource, Set[Acl]] = {
+    inReadLock(lock) {
+      aclCache.mapValues(_.acls).toMap
+    }
+  }
+
+  def close() {
+    if (aclChangeListener != null) aclChangeListener.close()
+    if (zkClient != null) zkClient.close()
+  }
+
+
+  private[auth] def startZkChangeListeners(): Unit = {
+    aclChangeListener = new ZkNodeChangeNotificationListener(zkClient, AclChangeNotificationZNode.path, AclChangeNotificationSequenceZNode.SequenceNumberPrefix, AclChangedNotificationHandler)
+    aclChangeListener.init()
+  }
+
+  private def updateCache(resource: Resource, versionedAcls: VersionedAcls) {
+    if (versionedAcls.acls.nonEmpty) {
+      aclCache.put(resource, versionedAcls)
```

Comment:
It looks like this needs to be write locked.

Reply:
good point. Fixing it.